src/NuGet.Core/NuGet.Packaging/Rules/ReferencesInNuspecMatchRefAssetsRule.cs

```diff
@@ -0,0 +1,137 @@
+            var missingItems = new List<MissingReference>();
+
+            foreach (var files in filesByTFM)
+            {
+                if(files.Key == "unsupported")
+                {
+                    continue;
+                }
+
+                var missingFiles = Array.Empty<string>();
+                var missingReferences = Array.Empty<string>();    
+                if (nuspecReferences.TryGetValue(files.Key, out var currentReferences))
+                {
+                    missingReferences = files.Where(m => !currentReferences.Contains(m)).ToArray();
+                    missingFiles = currentReferences.Where(t => !files.Contains(t)).ToArray();
+                }
+                else
+                {
+                    missingReferences = files.ToArray();
+                }
+
+                if (missingFiles.Length != 0)
+                {
+                    missingItems.Add(new MissingReference("ref", files.Key, missingFiles));
+                }
+
+                if (missingReferences.Length != 0)
+                {
+                    missingItems.Add(new MissingReference("nuspec", files.Key, missingReferences));
+                }
+            }
+            return missingItems;
+        }
+
+        internal IEnumerable<PackagingLogMessage> GenerateWarnings(IEnumerable<MissingReference> missingItems)
+        {
+            var issues = new List<PackagingLogMessage>();
+            if (missingItems.Count() != 0)
+            {
+                var message = new StringBuilder();
+                message.AppendLine(MessageFormat);
+                var referencesMissing = missingItems.Where(t => t.MissingFrom == "nuspec");
+                var refFilesMissing = missingItems.Where(t => t.MissingFrom == "ref");
+                foreach (var file in refFilesMissing)
+                {
+                    foreach(var item in file.MissingItems)
+                    {
+                        message.AppendLine(string.Format(_addToRefFormat, item, file.TFM));
+                    }
+                }
+
+                foreach (var reference in referencesMissing)
+                {
+                    foreach (var item in reference.MissingItems)
+                    {
+                        message.AppendLine(string.Format(_addToNuspecFormat, item, reference.TFM));
+                    }
+                }
+                issues.Add(PackagingLogMessage.CreateWarning(message.ToString(), NuGetLogCode.NU5131));
+            }
+            return issues;
+        }
+
+        private string GetFile(string filePath)
+        {
+            return filePath.Split('/')[filePath.Count(p => p == '/')];
+        }
+
+
+        internal class MissingReference
+        {
+            public string MissingFrom { get; }
+
+            public string TFM { get; }
```

Comment:
In .NET, acronyms are considered a word, and only the first letter of each word is caplitalised when using PascalCasing. For examples see the [Xml* classes in the System.Xml namespace](https://docs.microsoft.com/en-us/dotnet/api/system.xml?view=netframework-4.8), or the [Sql* classes in the System.Data.SqlClient namespace](https://docs.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlconnection?view=netframework-4.8)